### PR TITLE
Corral base use cases in Go Framework

### PR DESCRIPTION
**Pull request by slickwarren**
## Issue: 

https://github.com/rancher/qa-tasks/issues/357
 
## Problem

we want a way to rehydrate an rke2 cluster - specifically the local cluster running rancher
 
## Solution

In order to modify a local cluster, I added a test case for Corral. This also required modifying the dockerfile to install corral. The test suite currently can only handle running corrals that do not depend on eachother in any way, as the order is psudorandom ([see this issue](https://github.com/rancher/qa-tasks/issues/540) for more details)
In the test case, I include 2 corral packages: one to provision the local cluster (which was largely based on packges from the example https://github.com/rancherlabs/corral-packages repo) and another to modify it, replacing x number of server nodes in the cluster. Please see the updated readmes in their respective folder/package

While adding the test for corral. I also added some new variables when using corral in our golang framework. These are also updated in the readme, so please read it for more details (in standalone v2/validation test suite). IMO, the most important reason I added these is that the `corralConfigUser: <string, default="jenkauto">` user is what is used to name resources created in the backend, which is important for cleanup and correct naming practices we should be updating this to our own shortname (i.e. ctw for me). 
 
## Testing

local and remote testing (through jenkins) is done.  Will link privately.

### CODE REVIEW

File: tests/v2/validation/standalone/rke2/create_rancher_server/overlay/opt/corral/rke2/init-cluster.sh [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+mkdir -p /etc/rancher/rke2
+cat > /etc/rancher/rke2/config.yaml <<- EOF
+write-kubeconfig-mode: 644
+cni: ${CORRAL_cni}
+tls-san:
+  - ${CORRAL_api_host}
+EOF
+
+# apt install -y jq || true
+apt install -y jq
```
Comment (jakefhyde) [inline tests/v2/validation/standalone/rke2/create_rancher_server/overlay/opt/corral/rke2/init-cluster.sh:15]: This only works for distros using apt (e.g. ubuntu). We can also hash this out later (since this problem also exists in the upstream packages), but another solution may be to download jq directly from upstream, or use a distro agnostic package manager like linuxbrew.

File: tests/v2/validation/standalone/rke2/create_rancher_server/terraform/aws-rke2-rancher/aws-cluster-nodes/pools/main.tf [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+  }
+}
+
+resource "aws_instance" "agent" {
+  count = var.agent_count
+  ami = var.aws_ami
+  instance_type     = var.instance_type
+  key_name = aws_key_pair.corral_key.key_name
+  vpc_security_group_ids = [var.aws_security_group]
+  subnet_id = var.aws_subnet
```
Comment (jakefhyde) [inline tests/v2/validation/standalone/rke2/create_rancher_server/terraform/aws-rke2-rancher/aws-cluster-nodes/pools/main.tf:61]:
```suggestion
  count 								 = var.agent_count
  ami 									 = var.aws_ami
  instance_type     		 = var.instance_type
  key_name 							 = aws_key_pair.corral_key.key_name
  vpc_security_group_ids = [var.aws_security_group]
  subnet_id 						 = var.aws_subnet
```

File: tests/v2/validation/standalone/rke2/create_rancher_server/terraform/aws-rke2-rancher/aws-cluster-nodes/pools/main.tf [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+  name             = "${var.aws_hostname_prefix}-tg-80"
+  health_check {
+        protocol = "HTTP"
+        port = "traffic-port"
+        path = "/ping"
+        interval = 10
+        timeout = 6
+        healthy_threshold = 3
+        unhealthy_threshold = 3
+        matcher = "200-399"
```
Comment (jakefhyde) [inline tests/v2/validation/standalone/rke2/create_rancher_server/terraform/aws-rke2-rancher/aws-cluster-nodes/pools/main.tf:142]:
```suggestion
        protocol 						= "HTTP"
        port 								= "traffic-port"
        path 								= "/ping"
        interval 						= 10
        timeout 						= 6
        healthy_threshold 	= 3
        unhealthy_threshold = 3
        matcher 						= "200-399"
```

File: tests/v2/validation/standalone/rke2/create_rancher_server/terraform/aws-rke2-rancher/aws-cluster-nodes/pools/main.tf [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+  }
+}
+
+resource "aws_lb_target_group" "aws_tg_443" {
+  port             = 443
+  protocol         = "TCP"
+  vpc_id           = var.aws_vpc
+  name             = "${var.aws_hostname_prefix}-tg-443"
+  health_check {
+        protocol = "HTTP"
```
Comment (jakefhyde) [inline tests/v2/validation/standalone/rke2/create_rancher_server/terraform/aws-rke2-rancher/aws-cluster-nodes/pools/main.tf:152]: formatting

File: tests/framework/clients/corral/config.go [hunk 5]
```diff
@@ -13,11 +13,14 @@ const (
 // CorralConfigs is a struct that for necessary corral config environment variables to build a corral
 type CorralConfigs struct {
 	CorralConfigVars map[string]string `json:"corralConfigVars" yaml:"corralConfigVars"`
+	CorralConfigUser string            `json:"corralConfigUser" yaml:"corralConfigUser" default:"jenkauto"`
```
Comment (igomez06) [inline tests/framework/clients/corral/config.go:16]: What's the purpose for CorralConfigUser, is this a new corral feature.

File: tests/v2/validation/Dockerfile.validation [hunk 6]
```diff
@@ -17,6 +17,14 @@ COPY [".", "$WORKSPACE"]
 RUN go mod download && \
     go install gotest.tools/gotestsum@latest
 
+
+# Configure Corral
+ENV CORRAL_ROOT="/root/.corral"
+RUN chmod 600 .ssh/jenkins* && mkdir /root/.ssh && mkdir /root/.corral && ssh-keygen -f .ssh/jenkins* -y > /root/.ssh/public.pub
+RUN wget -O - https://github.com/rancherlabs/corral/releases/download/v1.1.1/corral-1.1.1-amd64 > corral && \
+    chmod +x corral && \
+    mv corral /usr/local/bin/
```
[the remark this answers is not in the document]
Comment (igomez06) [reply tests/v2/validation/Dockerfile.validation:26]: Jake wouldn't it be better to get the corral bin this way ?
`go install github.com/rancherlabs/corral@latest`

File: tests/v2/validation/standalone/corral_generic_test.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+const (
+	defaultRandStringLength = 5
+)
+
+type CorralStandaloneTestSuite struct {
+	suite.Suite
+	session *session.Session
+}
+
+func AppendRandomString(baseClusterName string) string {
```
[the remark this answers is not in the document]
Comment (igomez06) [reply tests/v2/validation/standalone/corral_generic_test.go:23]: Don't we already have this function?

File: tests/v2/validation/standalone/corral_generic_test.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+	corralConfig := corral.CorralConfigurations()
+	err := corral.SetupCorralConfig(corralConfig.CorralConfigVars, corralConfig.CorralConfigUser)
+	require.NoError(r.T(), err, "error reading corral configs")
+}
+
+func (r *CorralStandaloneTestSuite) TestGenericCorralPackage() {
+	corralPackage := corral.CorralPackagesConfig()
+	newPackages := []string{}
+	for key, value := range corralPackage.CorralPackagePath {
+		r.T().Log("key: ", key, "value: ", value)
```
[the remark this answers is not in the document]
Comment (igomez06) [reply tests/v2/validation/standalone/corral_generic_test.go:45]: Do we even need to log this? Seems unnecessary.

File: tests/v2/validation/standalone/rke2/create_rancher_server/manifest.yaml [hunk 9]
```diff
@@ -0,0 +1,128 @@
+name: aws-rke2-rancher-calico-v1.24.7-rke2r1-2.7.0
```
[the remark this answers is not in the document]
Comment (igomez06) [reply tests/v2/validation/standalone/rke2/create_rancher_server/manifest.yaml:1]: I don't get the purpose of this yaml, this is why we have corral-packages repo correct? We build the corral packages using that. That's how I have been doing it with code coverage.

File: tests/v2/validation/standalone/rke2/create_rancher_server/overlay/opt/corral/rke2/init-cluster.sh (hunk 1 above)
Comment (igomez06) [reply tests/v2/validation/standalone/rke2/create_rancher_server/overlay/opt/corral/rke2/init-cluster.sh:15]: Yes this exists in the upstream packages, definitely something we should look at. Using SLES it just ignores it as jq already exists on that distro

File: tests/v2/validation/Dockerfile.validation (hunk 6 above)
[the remark this answers is not in the document]
Comment (jakefhyde) [reply tests/v2/validation/Dockerfile.validation:26]: That's also an option, `latest` may produce surprising results though.

File: tests/v2/validation/standalone/corral_generic_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (jakefhyde) [reply tests/v2/validation/standalone/corral_generic_test.go:23]: I should've clarified, the function is already present in `/tests/v2/validation/provisioning/config.go`, we should move it to a util package instead since it's not strictly provisioning related.

File: tests/v2/validation/standalone/rke2/create_rancher_server/manifest.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (jakefhyde) [reply tests/v2/validation/standalone/rke2/create_rancher_server/manifest.yaml:1]: Generally yes, though I could see a use case where we would want to use/generate corral templates in rancher with a specific field, like the value of a current setting/env var, but I don't think we are quite there yet.

**Comment by slickwarren** [comment]
> I do not like having to copy over the corral package stuff from corral-packages to here at all. That is why we have that repo, it builds this stuff for us. 
I understand the concern with code bloat. I misunderstood the use of the corral-packages repo, and will refactor this PR accordingly.  

The changes I will make (as well as address other review comments):
* remove corral packages from rancher/rancher repo
* formatting updates
* update the corral_test to run from public corral images (built from CI in corral-packages repo, or built and deployed from local setups) instead of local packages

is there anything missing?

File: tests/v2/validation/Dockerfile.validation (hunk 6 above)
[the remark this answers is not in the document]
Comment (igomez06) [reply tests/v2/validation/Dockerfile.validation:26]: Or simply just `go install github.com/rancherlabs/corral@${CORRAL_VERSION}`

**Comment by igomez06** [comment]
> > I do not like having to copy over the corral package stuff from corral-packages to here at all. That is why we have that repo, it builds this stuff for us.
> 
> I understand the concern with code bloat. I misunderstood the use of the corral-packages repo, and will refactor this PR accordingly.
> 
> The changes I will make (as well as address other review comments):
> 
> * remove corral packages from rancher/rancher repo
> * formatting updates
> * update the corral_test to run from public corral images (built from CI in corral-packages repo, or built and deployed from local setups) instead of local packages
> 
> is there anything missing?

Thank you! I think that's it thank you Caleb!

File: tests/framework/clients/corral/config.go (hunk 5 above)
Comment (slickwarren) [reply tests/framework/clients/corral/config.go:16]: > IMO, the most important reason I added these is that the corralConfigUser: <string, default="jenkauto"> user is what is used to name resources created in the backend, which is important for cleanup and correct naming practices we should be updating this to our own shortname (i.e. ctw for me).

File: tests/v2/validation/Dockerfile.validation (hunk 6 above)
[the remark this answers is not in the document]
Comment (slickwarren) [reply tests/v2/validation/Dockerfile.validation:26]: I can use the go install option; does that require any of the `$PATH` or `chmod` work?

File: tests/v2/validation/standalone/corral_generic_test.go [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+		// if cleanup is enabled, delete packages that fail to correctly provision
+		if err != nil && corralPackage.Cleanup {
+			r.T().Log("Error Running Corral, attempting to clean up resources...  ")
+			corral.DeleteCorral(newPackageName)
+		}
+		require.NoError(r.T(), err, "error creating corral %v", key)
+		require.NotNil(r.T(), corralRun, "corral run had no output")
+	}
+	// if cleanup is enabled, delete packages in reverse order that they were created in
+	if corralPackage.Cleanup {
```
Comment (igomez06) [inline tests/v2/validation/standalone/corral_generic_test.go:59]: Why don't we use this Cleanup bool in conjunction with our session cleanup ie. when we create a corral we register it's delete function and when we defer session.Cleanup() it will also clean up the corral. Do you agree?

File: tests/framework/pkg/namegenerator/namegenerator.go [hunk 11]
```diff
@@ -38,3 +39,8 @@ func RandStringWithCharset(length int, charset string) string {
 func RandStringAll(length int) string {
 	return RandStringWithCharset(length, lowerLetterBytes+upperLetterBytes+numberBytes)
 }
+
+func AppendRandomString(baseClusterName string) string {
```
Comment (jakefhyde) [inline tests/framework/pkg/namegenerator/namegenerator.go:43]: :+1:

File: tests/framework/clients/corral/corral.go [hunk 12]
```diff
@@ -32,21 +37,22 @@ func SetupCorralConfig(configVars map[string]string) error {
...
-	} else {
-		msg, err := exec.Command("corral", "create", corralName, packagePath).CombinedOutput()
-		if err != nil {
-			return nil, errors.Wrap(err, "CreateCorral: "+string(msg))
-		}
+		args = append(args, debugFlag)
+	}
+	args = append(args, corralName, packageName)
+	msg, err := exec.Command("corral", args...).CombinedOutput()
+	myString = string(msg[:])
```
Comment (jakefhyde) [inline tests/framework/clients/corral/corral.go:51]: I'd probably rename `myString` to `output` or some other descriptive name

File: tests/v2/validation/standalone/corral_generic_test.go [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+	"github.com/rancher/rancher/tests/framework/clients/corral"
+
+	namegen "github.com/rancher/rancher/tests/framework/pkg/namegenerator"
+	"github.com/rancher/rancher/tests/framework/pkg/session"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+)
+
+const (
+	defaultRandStringLength = 5
```
Comment (jakefhyde) [inline tests/v2/validation/standalone/corral_generic_test.go:15]: I think this can be removed now

**Review by jakefhyde — COMMENTED** [review]
Few minor nits

**Comment by slickwarren** [comment]
This PR is ready for review again. There is 1 problem left that I will file a separate issue to track that I just couldn't figure out; how to log / fail the test when corral fails on a TF step.  For some reason, the go exec package does not capture TF output in the case of a failure (i.e. a bad AMI).  I've tried pulling the stderr and stdout pipes individually from the command, but ultimately I think it would need its own ticket as I suspect we will run into this when we have a terraform suite that Josh is working on

TLDR: everything should be addressed from the above comments, but corral failures won't fail the test with valid logs.  As mentioned, we will not be maintaining corral packages in this repo, so that was pulled from this PR and placed in the corral-packages repo.

**Review by jakefhyde — DISMISSED** [review]

File: tests/framework/clients/corral/corral.go [hunk 14]
```diff
@@ -30,24 +36,49 @@ func SetupCorralConfig(configVars map[string]string) error {
 	return nil
 }
 
+// SetCustomRepo sets a custom repo for corral to use. It takes a string as a parameter which is the repo you want to use
+func SetCustomRepo(repo string) error {
```
Comment (igomez06) [inline tests/framework/clients/corral/corral.go:40]: I'm concerned about this function. I always foresaw the corral-packages already cloned and built in a test environment. I'm not a big fan of this being a function. Like are we always going to clone and rebuild the repo when testing in Jenkins?

File: tests/v2/validation/provisioning/hosted/hosted_provisioning_test.go [hunk 15]
```diff
@@ -1,204 +0,0 @@
-package provisioning
```
Comment (igomez06) [inline tests/v2/validation/provisioning/hosted/hosted_provisioning_test.go:1]: I think your rebase is off, this shouldn't show up as a commit, I'm confused this is here.

File: tests/framework/clients/corral/corral.go (hunk 14 above)
Comment (slickwarren) [reply tests/framework/clients/corral/corral.go:40]: My understanding was that we would plan on publicly publishing the corral-packages repo at some point, but until then we need a solution that would still use a local build. Plus, this makes it easier to test corrals from other repos (i.e. your own fork) instead of relying on the package being in rancher's corral-packages repo.

Comment (igomez06) [reply tests/framework/clients/corral/corral.go:40]: Ok that makes sense, and just saw there is a config file flag for it too.

File: tests/framework/clients/corral/corral.go [hunk 16]
```diff
@@ -30,24 +36,49 @@ func SetupCorralConfig(configVars map[string]string) error {
...
-
-		myString := string(msg[:])
-		logrus.Infof(myString)
-	} else {
-		msg, err := exec.Command("corral", "create", corralName, packagePath).CombinedOutput()
-		if err != nil {
-			return nil, errors.Wrap(err, "CreateCorral: "+string(msg))
-		}
+func CreateCorral(ts *session.Session, corralName, packageName string, debug bool, cleanup bool) (*rest.Config, error) {
+	if cleanup {
```
Comment (igomez06) [inline tests/framework/clients/corral/corral.go:60]: Nice I definitely want this.

File: tests/framework/clients/corral/corral.go [hunk 17]
```diff
@@ -30,24 +36,49 @@ func SetupCorralConfig(configVars map[string]string) error {
...
+		args = append(args, debugFlag)
+	}
+	args = append(args, corralName, packageName)
+	logrus.Infof("Creating corral with the following parameters: %", args)
+	// complicated, but running the command in a way that allows us to
+	// capture the output and error(s) and print it to the console
+	msg, err := exec.Command("corral", args...).CombinedOutput()
+	logrus.Infof("Corral create output: %s", string(msg))
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to create corral: "+string(msg))
```
Comment (igomez06) [inline tests/framework/clients/corral/corral.go:80]: Did you figure out the issue with corral create not returning an error?

**Review by igomez06 — COMMENTED** [review]
Looks good, but something is off with the rebase since it's saying you have deleted a file that has already been deleted in the repo.

File: tests/framework/clients/corral/corral.go (hunk 17 above)
Comment (slickwarren) [reply tests/framework/clients/corral/corral.go:80]: [unfortunately not](https://github.com/rancher/rancher/pull/39752#issuecomment-1364263841).  However I believe the culprit here is Terraform, not necessarily corral.

File: tests/v2/validation/provisioning/hosted/hosted_provisioning_test.go (hunk 15 above)
Comment (slickwarren) [reply tests/v2/validation/provisioning/hosted/hosted_provisioning_test.go]: should be resolved now

**Review by igomez06 — APPROVED** [review]
LGTM

File: tests/framework/clients/corral/corral.go (hunk 14 above)
Comment (igomez06) [reply tests/framework/clients/corral/corral.go:40]: Caleb, I wanted to reopen this discussion as thought about this more. How necessary is that we would need to clone from repos from other forks? Back to your previous answer why we would need this function without publicly published corral-packages? Why couldn't we clone the repo beforehand in the JenkinsFile, or the DockerFile?

**Review by igomez06 — APPROVED** [review]
LGTM

**Review by jakefhyde — APPROVED** [review]
